contracts/LockedNORI.sol

```diff
@@ -40,127 +63,339 @@ contract LockedNORI is ERC777Upgradeable, ERC20PresetMinterPauserUpgradeable, ER
-        grant.claimedAmount;
+      if (grant.vestingSchedule.startTime > 0) {
+        // If any tokens have been revoked then the schedule (which doesn't get updated) could
+        // return more than the total grant amount.
+        return
+          MathUpgradeable.min(grant.grantAmount, grant.vestingSchedule.availableAmount(atTime)) - grant.claimedAmount;
+      } else {
+        return grant.grantAmount - grant.claimedAmount;
+      }
     }
-    return depositedBalanceOf(account);
+    return 0;
   }
 
+  /**
+   * @dev unlockedBalanceOf: Unlocked balance less any claimed amount
+   */
   function unlockedBalanceOf(address account) public view returns (uint256) {
     TokenGrant storage grant = grants[account];
     if (grant.exists) {
-      // Past the end date user can claim any remaining wrapped tokens
-      // TODO can't be larger than vested amount
+    //   console.log("grant.vestingSchedule.availableAmount(block.timestamp)",grant.vestingSchedule.availableAmount(block.timestamp));
+    //   console.log("grant.lockupSchedule.availableAmount(block.timestamp)",grant.lockupSchedule.availableAmount(block.timestamp));
+    //   console.log("grant.claimedAmount",grant.claimedAmount);
+      // If any tokens have been revoked then the schedule (which doesn't get updated) could
+      // return more than the total grant amount.
       return
-        grant.lockupSchedule.availableAmount(block.timestamp) -
-        grant.claimedAmount;
+        MathUpgradeable.min(
+          MathUpgradeable.min(
+            grant.vestingSchedule.availableAmount(block.timestamp),
+            grant.lockupSchedule.availableAmount(block.timestamp)
+          ),
+          grant.grantAmount
+        ) - grant.claimedAmount;
     }
-    return depositedBalanceOf(account);
+    return 0;
   }
 
-  function balanceOf(address account) public view override(ERC20Upgradeable, ERC777Upgradeable) returns (uint256) {
-    return unlockedBalanceOf(account);
+  /**
+   * @dev balanceOf: Total amount granted to the user less any previously claimed amount
+   */
+  function balanceOf(address account)
+    public
+    view
+    override(ERC20Upgradeable, ERC777Upgradeable)
+    returns (uint256)
+  {
+    return ERC777Upgradeable.balanceOf(account);
   }
 
-  function _beforeTokenTransfer(
-    address from,
-    address to,
-    uint256 amount
-  ) internal override(ERC20PresetMinterPauserUpgradeable, ERC20Upgradeable) {
-    if (to == address(0)) {
-      uint256 availableBalance = balanceOf(from);
-      uint256 wrappedBalance = super.balanceOf(from);
-      console.log(
-        "amount: %d  available: %d  wrapped: %d",
-        amount,
-        availableBalance,
-        wrappedBalance
-      );
-      require(amount <= availableBalance, "Withdrawl amount unavailable");
+  /**
+   * @dev getGrant: Returns all governing settings on a grant.
+   */
+  function getGrant(address account)
+    public
+    view
+    returns (
```

Comment:
I wonder if we should just return the struct?

Reply:
I defined a struct for the return type at least to document it better.